File: backend/internal/service/permission.go
package service

import (
	"keystonego/internal/model"
	"keystonego/internal/repository"

	"gorm.io/gorm"
)

// PermissionService 是权限定义管理业务逻辑服务。
type PermissionService struct {
	permRepo *repository.PermissionRepository
	db       *gorm.DB
}

// NewPermissionService 创建权限服务（Wire Provider）。
func NewPermissionService(permRepo *repository.PermissionRepository, db *gorm.DB) *PermissionService {
	return &PermissionService{permRepo: permRepo, db: db}
}

// PermissionListReq 权限列表查询请求参数。
type PermissionListReq struct {
	Keyword  string `form:"keyword"`   // 模糊搜索（匹配 path 和 description）
	Method   string `form:"method"`    // HTTP 方法过滤
	Status   *int8  `form:"status"`    // 状态过滤（nil 表示不过滤）
	Page     int    `form:"page"`      // 页码
	PageSize int    `form:"page_size"` // 每页条数
}

// List 多条件分页查询权限列表。
func (s *PermissionService) List(req PermissionListReq) ([]model.Permission, int64, error) {
	if req.Page <= 0 {
		req.Page = 1
	}
	if req.PageSize <= 0 {
		req.PageSize = 10
	}
	return s.permRepo.ListWithFilter(req.Keyword, req.Method, req.Status, req.Page, req.PageSize)
}

// GetByID 按 ID 查询权限定义。
func (s *PermissionService) GetByID(id uint64) (*model.Permission, error) {
	return s.permRepo.FindByID(id)
}

// Create 创建权限定义（默认状态启用）。
func (s *PermissionService) Create(perm *model.Permission) error {
	perm.Status = 1
	return s.permRepo.Create(perm)
}

// Update 更新权限定义（部分更新）。
func (s *PermissionService) Update(id uint64, perm *model.Permission) error {
	existing, err := s.permRepo.FindByID(id)
	if err != nil {
		return err
	}
	if perm.Path != "" {
		existing.Path = perm.Path
	}
	if perm.Method != "" {
		existing.Method = perm.Method
	}
	if perm.Description != "" {
		existing.Description = perm.Description
	}
	existing.Status = perm.Status
	return s.permRepo.Update(existing)
}

// Delete 删除权限定义。
func (s *PermissionService) Delete(id uint64) error {
	return s.permRepo.Delete(id)
}

// InitDefaultPermissions 初始化默认权限定义（仅在表为空时执行，保证幂等）。
func (s *PermissionService) InitDefaultPermissions() error {
	var count int64
	s.db.Model(&model.Permission{}).Count(&count)
	if count > 0 {
		return nil // 已有数据，跳过初始化
	}

	// 默认权限定义：各资源的 CRUD 操作
	perms := []model.Permission{
		{Path: "/api/v1/users", Method: "GET", Description: "查询用户列表"},
		{Path: "/api/v1/users", Method: "POST", Description: "创建用户"},
		{Path: "/api/v1/users", Method: "PUT", Description: "更新用户"},
		{Path: "/api/v1/users", Method: "DELETE", Description: "删除用户"},
		{Path: "/api/v1/roles", Method: "GET", Description: "查询角色列表"},
		{Path: "/api/v1/roles", Method: "POST", Description: "创建角色"},
		{Path: "/api/v1/roles", Method: "PUT", Description: "更新角色"},
		{Path: "/api/v1/roles", Method: "DELETE", Description: "删除角色"},
		{Path: "/api/v1/menus", Method: "GET", Description: "查询菜单列表"},
		{Path: "/api/v1/menus", Method: "POST", Description: "创建菜单"},
		{Path: "/api/v1/menus", Method: "PUT", Description: "更新菜单"},
		{Path: "/api/v1/menus", Method: "DELETE", Description: "删除菜单"},
		{Path: "/api/v1/permissions", Method: "GET", Description: "查询权限列表"},
		{Path: "/api/v1/permissions", Method: "POST", Description: "创建权限"},
		{Path: "/api/v1/permissions", Method: "PUT", Description: "更新权限"},
		{Path: "/api/v1/permissions", Method: "DELETE", Description: "删除权限"},
	}

	for _, p := range perms {
		s.permRepo.Create(&p)
	}
	return nil
}

File: frontend/src/api/permission.ts
/**
 * 权限 API 模块 — 权限定义的 CRUD。
 *
 * Permission 对应后端 Permission 表（权限定义），与 Casbin 策略是不同维度：
 *   - 权限定义：前端展示用，描述有哪些 API 端点可被授权
 *   - Casbin 策略：运行时鉴权用，p 规则定义角色-路径-方法的映射
 */
import request from '@/utils/request'

/** 权限定义接口 */
export interface Permission {
  id: number
  path: string       // API 路径，如 /api/v1/users
  method: string     // HTTP 方法，如 GET/POST/PUT/DELETE
  description: string // 权限描述
  status: number     // 1=启用, 0=禁用
  created_at: string
  updated_at: string
}

/** 分页查询权限列表（支持关键词和方法过滤） */
export function getPermissions(params: { page?: number; page_size?: number; keyword?: string; method?: string }) {
  return request.get('/permissions', { params })
}

/** 按 ID 查询权限 */
export function getPermission(id: number) {
  return request.get(`/permissions/${id}`)
}

/** 创建权限定义 */
export function createPermission(data: Partial<Permission>) {
  return request.post('/permissions', data)
}

/** 更新权限定义 */
export function updatePermission(id: number, data: Partial<Permission>) {
  return request.put(`/permissions/${id}`, data)
}

/** 删除权限定义 */
export function deletePermission(id: number) {
  return request.delete(`/permissions/${id}`)
}
